Refactor Time and Sleep - MainLoop 2 of N

fmatthew5876 opened the pull request:
Depends on #2021 

This adds `Game_Clock` for getting the current time and and sleeping. 

- [x] Remove dependency on Ui for time keeping
- [x] Use highest precision time available on system
- [x] Prefer standard time functions over platform specific ones wherever possible

TBD:
- [x] Check if we can use `std::this_thread::sleep_for()` on 3ds
- [x] Check if we can use `std::this_thread::sleep_for()` on switch
- [x] Check if we can use `std::this_thread::sleep_for()` on vita

fmatthew5876 commented:
Is there any known platform where `std::this_thread::sleep_for()` doesn't work? I just added a commit here to get rid of our `Sleep()` functions which I hope we can keep.

Ghabry commented:
> Is there any known platform where std::this_thread::sleep_for() doesn't work?

Hard to answer, would need a test on Wii, Vita, Switch and 3DS. No idea how this is implemented when using newlib, could be a stub or working.

fmatthew5876 commented:
This is getting folded into #2017

fmatthew5876 commented:
I moved the clock code from #2017 here. Rebased everything and now this is ready.

For now I have kept the platform specific sleep functions for 3ds, switch, and psp2. Would love to know if we can get rid of these but that can also be done later.

Ghabry commented on src/game_clock.cpp:
```diff
@@ -17,5 +17,28 @@
+	auto us = std::chrono::duration_cast<std::chrono::microseconds>(ns);
+	sceKernelDelayThread(us.count());
+#else
+	std::this_thread::sleep_for(ns);
```
based on the define of _GLIBCXX_USE_NANOSLEEP (they all use libstdc++) except for Vita all support ::nanosleep (which is called by sleep_for)

carstene1ns commented on src/switch_ui.cpp:
```diff
@@ -326,21 +326,10 @@ NxUi::~NxUi() {
-	svcSleepThread(nsecs);
-}
-
 static inline double u64_to_double(u64 value) {
```
This one is useless now.

carstene1ns commented on src/3ds_ui.cpp:
```diff
@@ -147,20 +147,10 @@ CtrUi::~CtrUi() {
-	svcSleepThread(nsecs);
-}
-
 static inline double u64_to_double(u64 value) {
```
same here

carstene1ns commented on src/game_clock.h:
```diff
@@ -60,4 +69,18 @@ inline uint32_t Game_Clock::GetTicks() {
+	return DEFAULT_FPS;
+}
+
+/** Get the amount of time each logical frame should take */
```
duplicated comment

fmatthew5876 commented:
Rebased and addressed review comments
